proto-lens/src/Data/ProtoLens/Encoding.hs

```diff
@@ -137,42 +157,51 @@ buildMessage msg = foldMap putTaggedValue (messageToTaggedValues msg)
 -- | Encode a message as a sequence of key-value pairs.
 messageToTaggedValues :: Message msg => msg -> [TaggedValue]
 messageToTaggedValues msg = mconcat
-    [ map (TaggedValue t) (messageFieldToVals fieldDescr msg)
+    [ messageFieldToVals t fieldDescr msg
     | (Tag t, fieldDescr) <- Map.toList (fieldsByTag descriptor)
     ]
 
-messageFieldToVals :: FieldDescriptor msg -> msg -> [WireValue]
-messageFieldToVals (FieldDescriptor _ typeDescriptor accessor) msg =
-    case fieldWireType typeDescriptor of
-        FieldWireType wt convert _ -> case accessor of
+messageFieldToVals :: Int -> FieldDescriptor a -> a -> [TaggedValue]
+messageFieldToVals tag (FieldDescriptor _ typeDescriptor accessor) msg =
+    let
+        embed src
+            = case fieldWireType typeDescriptor of
+                FieldWireType wt convert _
+                    -> [TaggedValue tag $ WireValue wt
+                                            (convert src)]
+                GroupFieldType
+                    -> TaggedValue tag (WireValue StartGroup ())
+                            : messageToTaggedValues src
+                                ++ [TaggedValue tag $ WireValue EndGroup ()]
```

Comment:
Could be faster by changing messageFieldToVals to take a "rest" parameter (DList-style).  Maybe not worth the complexity though, since groups are deprecated.